File: mq/rabbitmq/py/broadcast/subscriber_saver.py
#!/usr/bin/env python3
#
# Copyright (C) 2019 jayson
#
import json
import logging
import contextlib
import utils.logging
from utils.connection import get_exchange_connection
from config import RABBITMQ_MASTER
from pika.adapters.blocking_connection import BlockingChannel
from pika.spec import Basic, BasicProperties


@contextlib.contextmanager
def create_consumer(filename: str):
    with open(filename, 'a') as fp:
        def __consumer(ch: BlockingChannel, method: Basic.Deliver, properties: BasicProperties, body: bytes):
            data = json.loads(body.decode('utf-8'))
            utils.logging.info('{}@{}@{}: {}'.format(data['from'], data['producer'], data['index'], data['body']),
                               file=fp)
            ch.basic_ack(delivery_tag=method.delivery_tag)
        yield __consumer


if __name__ == '__main__':
    utils.logging.init_logging(logging.DEBUG)
    with get_exchange_connection(RABBITMQ_MASTER, 'logs_broadcast', 'fanout') as channel:
        channel.queue_declare(queue='logs_broadcast_saver', durable=True, auto_delete=True)
        channel.queue_bind(queue='logs_broadcast_saver', exchange='logs_broadcast')

        with create_consumer('broadcast.log') as consumer:
            channel.basic_consume(queue='logs_broadcast_saver', on_message_callback=consumer)
            channel.start_consuming()

File: mq/rabbitmq/php/src/broadcast/subscriber_saver.php
<?php
/**
 * rabbitmq hello-world
 */
require_once __DIR__ . '/../bootstrap.php';


$connection = new \PhpAmqpLib\Connection\AMQPStreamConnection(RABBITMQ_MASTER, 5672, 'guest', 'guest');
$channel = $connection->channel();

$channel->queue_declare('logs_broadcast_saver', false, true, false, true);
$channel->queue_bind('logs_broadcast_saver', 'logs_broadcast', 'logs_broadcast_saver');

$channel->basic_consume('logs_broadcast_saver', null, false, false, false, false,
    function(\PhpAmqpLib\Message\AMQPMessage $message) {
        $data = json_decode($message->body);
        file_put_contents('broadcast.log', sprintf("<%s:%s@%s>: %s\n", $data->from, $data->producer, $data->index, $data->body), FILE_APPEND);

        /* @var \PhpAmqpLib\Channel\AMQPChannel $channel */
        $channel = $message->delivery_info['channel'];
        return $channel->basic_ack($message->getDeliveryTag());
    }
);

while ($channel->is_consuming()) {
    $channel->wait();
}
